Add CAN bus and ODrive Motor components

Slaith12 opened the pull request:
# Feature
* Adds can_bus and odrive_motor components
* can_bus handles all CAN messaging and bus configuration, and is able to pass on received messages to other components that call add_listener()
* odrive_motor controls the ODrive motor controller and can be used to read its current state and control its position/velocity

# How to Test
* Run odrive_motor_test.py

Slaith12 committed:
Created can bus & ODrive components

Slaith12 committed:
odrive improvements

Slaith12 committed:
Add test script

Slaith12 committed:
update test code

mdrc-catbot committed:
Update ODrive Code

hiromon0125 commented on src/raspi/component/can/can_bus.py:
```diff
@@ -1,64 +1,64 @@
+)
+
+@device
+@dataclass(slots = True)
```
theres been an issue with slots so you can just delete that

hiromon0125 commented on src/raspi/component/motor/odrive_motor.py:
```diff
@@ -0,0 +1,125 @@
+    identifier
+)
+
+db = cantools.database.load_file("src/raspi/odrive-cansimple.dbc")
```
maybe move the dbc to the motor directory

hiromon0125 reviewed (commented):
lgtm, lets just get this fully tested today and make sure this all works

Slaith12 committed:
Update ODrive test code

Slaith12 committed:
Add more detail to test screen

Slaith12 committed:
Update odrive_motor.py

mdrc-catbot committed:
Finish test code

hiromon0125 commented on src/raspi/component/can/can_bus.py:
```diff
@@ -1,64 +1,64 @@
-    axis_id: int
-    can: CanBus
+        self.listeners = {}
+        self.notifier = can.Notifier(self.bus, [lambda msg: read_message(self, msg)])
```
```py
DEFAULT_BUS_CONFIG = {"channel": "can0", "interface": "socketcan", "config_context": None, ignore_config: False}

class CanBus:
    busConfig: dict = DEFAULT_BUS_CONFIG
    bus: can.Bus = None
    notifier: can.Notifier = None
    listeners = field(default_factory=dict)

    def __post_init__(self):
        self.busConfig = {**DEFAULT_BUS_CONFIG, **self.busConfig}
        self.bus = self.bus if self.bus else can.Bus(**self.busConfig) 
        self.notifier = self.notifier if self.notifier else can.Notifier(self.bus, (lambda msg: read_message(self, msg),))
```
I'd recommend this implementation for several reasons:
1. While we probably will not have more than one CanBus object, there will be errors of multiple CanBus objects all sharing the same listeners dictionary object, so `field(default_factory=dict)` will fix this bug and is a good practice to begin with. (https://docs.python.org/3/library/dataclasses.html#mutable-default-values)
2. Debugging with __repr__. You will be able to see all of the config params when converted to a string
3. Able to dependency inject during testing

hiromon0125 commented on src/raspi/component/can/can_bus.py:
```diff
@@ -1,64 +1,64 @@
+    if axisId in can_bus.listeners:
+        can_bus.listeners[axisId](msg)
+    else:
+        logging.info("Unhandled message with id " + hex(msg.arbitration_id))
```
You can log info regardless of handle or not, and log warn if it was not handled.

hiromon0125 commented on src/raspi/component/can/can_bus.py:
```diff
@@ -1,64 +1,64 @@
+        bus.bus.send(msg)
+        return True
+    except can.CanError:
+        return False
```
log the error here

hiromon0125 commented on src/raspi/component/can/can_bus.py:
```diff
@@ -1,64 +1,64 @@
 
+@device_action(ctx)
+def close(bus: CanBus) -> None:
+    bus.notifier.stop()
```
you can log info shutting down canBus

hiromon0125 commented on src/raspi/component/motor/odrive_motor.py:
```diff
@@ -0,0 +1,136 @@
+db = cantools.database.load_file("src/raspi/component/motor/odrive-cansimple.dbc")
+
+@device
+@dataclass()
```
```suggestion
@dataclass
```

hiromon0125 commented on src/raspi/component/motor/odrive_motor.py:
```diff
@@ -0,0 +1,136 @@
+        case "Get_Controller_Error": #0x1D
+            logging.error("Controller Error w/ following data: " + msg_db.decode(msg.data))
+        case _:
+            logging.warning(f"Unrecognized message \"{msg_db.name}\" recieved (id {hex(msg_id)})")
```
```suggestion
EVENT_HANDLE = dict()
EVENT_IGNORE = ["MotorStatus"]


def event_handle(motor, msg):
    msg_id = msg.arbitration_id & 0x1F
    msg_db = db.get_message_by_frame_id(msg_id)
    if msg_db.name in EVENT_HANDLE:
        EVENT_HANDLE[msg_db.name](motor, msg)
    elif msg_db.name not in EVENT_IGNORE:
        logging.warning(
            f'Unrecognized message "{msg_db.name}" recieved (id {hex(msg_id)})'
        )


def event_decorator(name):
    def wrapper(func):
        EVENT_HANDLE[name] = func
        return func

    return wrapper


@event_decorator("Heartbeat")
def motor_status(motor, msg):
    # Do something with the message
    pass


@event_decorator("Get_Motor_Status")
def motor_status(motor, msg):
    # Do something with the message
    pass

```
This is a more readable syntax for declaring this instead of a 40 line function with all of the event edgecases.

hiromon0125 commented on src/raspi/component/motor/odrive_motor.py:
```diff
@@ -0,0 +1,136 @@
+@device_parser(ctx)
+def parse_odrive(data: dict) -> ODriveMotor:
+    motor = ODriveMotor(**data)
+    return motor
```
```suggestion
    return ODriveMotor(**data)
```

hiromon0125 commented on src/raspi/component/motor/odrive_motor.py:
```diff
@@ -0,0 +1,136 @@
+
+@device_action(ctx)
+def calibrate(motor) -> bool:
+    pass
```
delete

hiromon0125 commented on src/raspi/component/motor/odrive_motor.py:
```diff
@@ -0,0 +1,136 @@
+    identifier
+)
+
+db = cantools.database.load_file("src/raspi/component/motor/odrive-cansimple.dbc")
```
```suggestion
ODRIVE_CAN_DB = cantools.database.load_file("src/raspi/component/motor/odrive-cansimple.dbc")
```

hiromon0125 reviewed (changes requested):
see comments

hiromon0125 commented:
Add a little bit more detailed description of the features added in this pr

Slaith12 committed:
Fixes and cleanup

Slaith12 committed:
Update odrive_motor.py

mdrc-catbot committed:
Extra fixes

Slaith12 committed:
Add better position control & adjust test code

Code now learns what control/input mode the motor starts out in. Also added optional parameters for position limits.

mdrc-catbot committed:
Update position control code

Slaith12 committed:
Add trajectory control for ODrive motors
